### PR TITLE
Prod: metrics, alerts, kpis

### PR DESCRIPTION
**Pull request by Ishankoradia**
## Summary by CodeRabbit

* **New Features**
  * Metrics CRUD (column or SQL-expression), payload validation, preview values, and consumer lookup for charts/KPIs.
  * KPI CRUD tied to metrics with targets/thresholds, trend charts, summaries, and dashboard KPI components.
  * KPI annotations (notes) and report snapshot KPI data endpoint.
* **Permissions**
  * New permissions for viewing/creating/editing/deleting Metrics and KPIs.
* **Behavior**
  * Comments/mentions now use a unified target_id and support KPIs; previews indicate when a warehouse is not configured.

[Review Change Stack](https://app.coderabbit.ai/change-stack/DalgoT4D/DDP_backend/pull/1312?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: ddpui/core/metric/metric_service.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
+"""Metric service for business logic
```
Comment (siddhant3030) [inline ddpui/core/metric/metric_service.py:1]: ### Services placed in wrong directory

Per `CLAUDE.md`, the architecture prescribes:
- `ddpui/core/{module}/{module}_service.py` for business logic
- *"Services folder removed — merged into core with feature-based organization"*

**Suggestion:** Move to `ddpui/core/metric/metric_service.py` with this structure:
```
ddpui/core/metric/
├── __init__.py              # Keep empty
├── metric_service.py        # This file
└── exceptions.py            # Extract exceptions here
```

Same applies to `kpi_service.py` → `ddpui/core/kpi/kpi_service.py`.

File: ddpui/services/metric_service.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+
+logger = CustomLogger("ddpui.metric_service")
+
+VALID_AGGREGATIONS = [choice[0] for choice in AGGREGATION_CHOICES]
+
+
+# ── Exceptions ──────────────────────────────────────────────────────────────
+
+
+class MetricServiceError(Exception):
```
Comment (siddhant3030) [inline ddpui/services/metric_service.py]: ### Exceptions should be in a separate file

Per `CLAUDE.md`: exceptions belong in `ddpui/core/{module}/exceptions.py`, not inline in the service file.

**Suggestion:** Extract `MetricServiceError`, `MetricNotFoundError`, `MetricValidationError`, and `MetricDeleteBlockedError` into `ddpui/core/metric/exceptions.py`.

File: ddpui/services/metric_service.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    @staticmethod
+    def validate_metric_against_warehouse(metric: Metric, org_warehouse: OrgWarehouse) -> None:
+        """Execute a test query against the warehouse to validate the metric definition."""
+        try:
+            warehouse = WarehouseFactory.get_warehouse_client(org_warehouse)
+            qb = AggQueryBuilder()
+            qb.fetch_from(metric.table_name, metric.schema_name)
+
+            if metric.column_expression:
+                qb.add_column(literal_column(metric.column_expression).label("metric_value"))
```
Comment (siddhant3030) [inline ddpui/services/metric_service.py]: ### SQL injection risk via `column_expression`

`literal_column()` passes the expression directly into SQL **without any escaping or parameterization**. This is used in 5 places across the codebase. Any user with `can_create_metrics` permission can inject arbitrary SQL — e.g., `1; DROP TABLE users --` or data exfiltration queries.

The `validate_metric_against_warehouse` check only validates syntax (the malicious SQL would execute successfully during validation too).

**Suggestion — pick one or combine:**
1. **Allowlist approach**: Parse the expression and only permit known SQL functions (`SUM`, `COUNT`, `AVG`, `MIN`, `MAX`, `COALESCE`, arithmetic operators). Reject anything else.
2. **Sandboxed validation**: Execute the test query inside a read-only transaction / role with no DDL permissions.
3. **Restrict permissions**: At minimum, ensure `can_create_metrics` and `can_edit_metrics` are only granted to admin/account-manager roles (verify in seed data). Document this as a known risk.
4. **Regex guard**: Block obvious patterns like `;`, `DROP`, `DELETE`, `INSERT`, `UPDATE`, `ALTER`, `GRANT`, `--`, `/*` in the expression before passing to `literal_column()`.

File: ddpui/services/metric_service.py [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+            return float(value) if value is not None else None
+        return None
+
+    @staticmethod
+    def get_metric_consumers(metric_id: int, org: Org) -> dict:
+        """Find charts and KPIs that reference this metric."""
+        # Verify metric exists
+        MetricService.get_metric(metric_id, org)
+
+        from ddpui.models.metric import KPI
```
Comment (siddhant3030) [inline ddpui/services/metric_service.py]: ### Local import inside function

Per `CLAUDE.md`: *"No Local Imports: Always use global imports at the top of the file. Never use local/inline imports inside functions or methods."*

**Suggestion:** Move `from ddpui.models.metric import KPI` to the top of the file with the other imports. `KPI` is already in the same `ddpui.models.metric` module that's imported at line 12, so you can just add it there:
```python
from ddpui.models.metric import Metric, KPI, AGGREGATION_CHOICES
```

File: ddpui/core/metric/metric_service.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+        MetricService.get_metric(metric_id, org)
+
+        from ddpui.models.metric import KPI
+
+        # KPIs via FK
+        kpis = list(KPI.objects.filter(metric_id=metric_id, org=org).values("id", "name"))
+
+        # Charts via saved_metric_id in extra_config
+        charts = []
+        for chart in Chart.objects.filter(org=org):
```
Comment (siddhant3030) [inline ddpui/core/metric/metric_service.py:339]: ### 💡 Suggestion: Chart consumer scan loads all charts into memory

`Chart.objects.filter(org=org)` loads every chart for the org and iterates through JSON one-by-one. Fine at current scale (~20 NGOs), but this is O(n) per call.

**Suggestion:** If using PostgreSQL, you can use a JSON lookup:
```python
from django.db.models import Q
charts = Chart.objects.filter(
    org=org,
    extra_config__metrics__contains=[{"saved_metric_id": metric_id}]
)
```
Or add a `ManyToManyField` / through table for saved metric references on `Chart` for efficient lookups.

File: ddpui/api/metric_api.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    MetricDeleteBlockedError,
+)
+from ddpui.utils.custom_logger import CustomLogger
+
+logger = CustomLogger("ddpui")
+
+metric_router = Router()
+
+
+@metric_router.get("/", response=MetricListResponse)
```
Comment (siddhant3030) [inline ddpui/api/metric_api.py:33]: ### ❌ Blocking: Missing `api_response()` wrapper and `ApiResponse` response type

Per `CLAUDE.md`: *"Always wrap responses with `api_response()`"* and use `response=ApiResponse[SchemaType]`.

All endpoints in this file return schema objects directly without wrapping.

**Suggestion:** Follow the template from CLAUDE.md:
```python
from ddpui.utils.response_wrapper import api_response, ApiResponse

@metric_router.get("/", response=ApiResponse[MetricListResponse])
@has_permission(["can_view_metrics"])
def list_metrics(request, ...):
    ...
    return api_response(
        success=True,
        data=MetricListResponse(...)
    )
```

Also applies to `delete_metric` (line 195) which returns a raw `{"success": True}` dict — should use `api_response(success=True, message="Metric deleted successfully")`.

Same issue in `kpi_api.py`.

File: ddpui/api/metric_api.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        page=page,
+        page_size=page_size,
+        search=search,
+        schema_name=schema_name,
+        table_name=table_name,
+    )
+
+    total_pages = (total + page_size - 1) // page_size
+
+    metric_responses = [
```
Comment (siddhant3030) [inline ddpui/api/metric_api.py:62]: ### 💡 Suggestion: Add `from_model()` to schemas to eliminate repetitive field-by-field construction

This manual `MetricResponse(id=m.id, name=m.name, ...)` pattern is repeated 4 times in this file alone (lines 61-72, 105-116, 130-141, 168-179).

Per `CLAUDE.md`: *"Include `from_model()` class methods for easy conversion"*.

**Suggestion:** Add to `MetricResponse`:
```python
@classmethod
def from_model(cls, m: "Metric") -> "MetricResponse":
    return cls(
        id=m.id, name=m.name, description=m.description,
        schema_name=m.schema_name, table_name=m.table_name,
        column=m.column, aggregation=m.aggregation,
        column_expression=m.column_expression,
        created_at=m.created_at, updated_at=m.updated_at,
    )
```
Then all 4 call sites become `MetricResponse.from_model(m)`. Same for `KPIResponse`.

File: ddpui/api/metric_api.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+)
+from ddpui.services.metric_service import (
+    MetricService,
+    MetricNotFoundError,
+    MetricValidationError,
+    MetricDeleteBlockedError,
+)
+from ddpui.utils.custom_logger import CustomLogger
+
+logger = CustomLogger("ddpui")
```
Comment (siddhant3030) [inline ddpui/api/metric_api.py:28]: ### 🔍 Nitpick: Logger name should include module

Per CLAUDE.md template: `CustomLogger("ddpui.{module}_api")`.

**Suggestion:** `CustomLogger("ddpui.metric_api")` instead of `CustomLogger("ddpui")`.

Same in `kpi_api.py:24`.

File: ddpui/core/charts/charts_service.py [hunk 9]
```diff
@@ -162,6 +162,52 @@ def normalize_dimensions(payload: ChartDataPayload) -> List[str]:
...
+        if not isinstance(m, dict):
+            resolved.append(m)
+            continue
+
+        saved_id = m.get("saved_metric_id")
+        if saved_id is None:
+            resolved.append(m)
+            continue
+
+        from ddpui.models.metric import Metric
```
Comment (siddhant3030) [inline ddpui/core/charts/charts_service.py]: ### ❌ Blocking: Local import inside function

Per `CLAUDE.md`: *"No Local Imports: Always use global imports at the top of the file."*

**Suggestion:** Move `from ddpui.models.metric import Metric` to the top of the file with the other imports.

File: ddpui/services/kpi_service.py [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+            return "green"
+        if achievement <= amber_pct:
+            return "amber"
+        return "red"
+
+
+# ── Exceptions ──────────────────────────────────────────────────────────────
+
+
+class KPIServiceError(Exception):
```
Comment (siddhant3030) [inline ddpui/services/kpi_service.py]: ### ❌ Blocking: Exceptions should be in a separate file

Same as metric_service — per `CLAUDE.md`, exceptions belong in `ddpui/core/kpi/exceptions.py`.

**Suggestion:** Extract `KPIServiceError`, `KPINotFoundError`, `KPIValidationError` into their own file.

File: ddpui/services/kpi_service.py [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+
+        return results
+
+    @staticmethod
+    def get_kpi_data(kpi_id: int, org: Org) -> dict:
+        """Get KPI chart data + echarts config.
+
+        Same pattern as chart data endpoint: returns {data, echarts_config}.
+        """
+        from ddpui.core.charts.echarts_config_generator import EChartsConfigGenerator
```
Comment (siddhant3030) [inline ddpui/services/kpi_service.py]: ### ❌ Blocking: Local import inside function

Per `CLAUDE.md`: *"No Local Imports: Always use global imports at the top of the file."*

**Suggestion:** Move `from ddpui.core.charts.echarts_config_generator import EChartsConfigGenerator` to the top of the file.

File: ddpui/services/kpi_service.py [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+    if target_value == 0:
+        return None
+    achievement = (current_value / target_value) * 100
+    if direction == "increase":
+        if achievement >= green_pct:
+            return "green"
+        if achievement >= amber_pct:
+            return "amber"
+        return "red"
+    else:  # decrease — lower is better
```
Comment (siddhant3030) [inline ddpui/services/kpi_service.py]: ### ❌ Blocking: RAG logic bug with default thresholds for "decrease" direction

With default thresholds (`green=100.0`, `amber=80.0`) and `direction=decrease`:
- `achievement <= 100` → green (catches almost everything at or below target)
- `achievement <= 80` → amber (**unreachable** because green catches it first)
- else → red

There is **no amber zone** with default thresholds when direction is decrease. An NGO user creating a "reduce dropout rate" KPI with defaults would never see amber — only green or red.

**Suggestion — pick one:**
1. **Validate thresholds against direction** in `_validate_fields()`: require `amber_pct > green_pct` when `direction=decrease` (and `amber_pct < green_pct` when `direction=increase`).
2. **Swap defaults based on direction**: when `direction=decrease`, use `green_threshold_pct=100.0, amber_threshold_pct=120.0` as defaults.
3. **At minimum**, add a warning log if `amber_pct < green_pct` for decrease direction.

File: ddpui/schemas/kpi_schema.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+
+# ── KPI Schemas ─────────────────────────────────────────────────────────────
+
+
+class KPICreate(Schema):
+    metric_id: int
+    name: Optional[str] = None  # defaults to metric name
+    target_value: Optional[float] = None
+    direction: str  # "increase" or "decrease"
```
Comment (siddhant3030) [inline ddpui/schemas/kpi_schema.py:17]: ### 💡 Suggestion: Validate enum fields at schema level

Per `CLAUDE.md`: *"Schema Validation: Validate data using schemas before any database/external operations."*

`direction` and `time_grain` accept any string and are only validated deep in the service layer.

**Suggestion:** Use `Literal` types for compile-time safety and immediate 422 validation:
```python
from typing import Literal

direction: Literal["increase", "decrease"]
time_grain: Literal["daily", "weekly", "monthly", "quarterly", "yearly"]
metric_type_tag: Optional[Literal["input", "output", "outcome", "impact"]] = None
```

Same for `MetricCreate.aggregation` in `metric_schema.py:21`.

File: ddpui/schemas/metric_schema.py [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    """Response for metric value preview"""
+
+    value: Optional[float] = None
+    error: Optional[str] = None
+
+
+class MetricConsumersResponse(Schema):
+    """Response for metric consumers (charts and KPIs referencing this metric)"""
+
+    charts: List[dict]  # [{id, title, chart_type}]
```
Comment (siddhant3030) [inline ddpui/schemas/metric_schema.py:61]: ### 💡 Suggestion: Use typed schemas instead of `List[dict]`

`List[dict]` provides no type safety or API documentation.

**Suggestion:**
```python
class ChartConsumerRef(Schema):
    id: int
    title: str
    chart_type: str

class KPIConsumerRef(Schema):
    id: int
    name: str

class MetricConsumersResponse(Schema):
    charts: List[ChartConsumerRef]
    kpis: List[KPIConsumerRef]
```

File: ddpui/models/metric.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+    # Option B: Expression — free-text expression
+    # e.g. "SUM(col_a - col_b) / COUNT(DISTINCT id)"
+    # Validated on save by executing a test query against the warehouse
+    column_expression = models.TextField(null=True, blank=True)
+
+    # Org scoping + ownership (same pattern as Chart)
+    org = models.ForeignKey(Org, on_delete=models.CASCADE)
+    created_by = models.ForeignKey(
+        OrgUser, on_delete=models.CASCADE, related_name="metrics_created"
+    )
```
Comment (siddhant3030) [inline ddpui/models/metric.py:64]: ### 💡 Suggestion: Add `last_modified_by` field

Both `Metric` and `KPI` track `created_by` but not who last modified the record. The existing `Chart` model in the codebase includes `last_modified_by` — these should follow the same pattern for audit trail consistency.

**Suggestion:**
```python
last_modified_by = models.ForeignKey(
    OrgUser, on_delete=models.CASCADE, related_name="metrics_modified", null=True
)
```
Then set it in `update_metric()` and `update_kpi()`.

File: ddpui/api/kpi_api.py [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+    )
+
+
+@kpi_router.get("/", response=KPIListResponse)
+@has_permission(["can_view_kpis"])
+def list_kpis(
+    request,
+    page: int = 1,
+    page_size: int = 10,
+    search: str = None,
```
Comment (siddhant3030) [inline ddpui/api/kpi_api.py]: ### 🔍 Nitpick: `search: str = None` should be `Optional[str] = None`

For type correctness. Same for `program_tag` and `metric_type` below.

Same issue in `metric_api.py:37-39`.

**Review by siddhant3030 — CHANGES_REQUESTED** [review]
## PR Review: metrics, alerts, kpis

**PR**: #1312 (`prod/metrics_kpis_alerts` → `main`)
**Files Changed**: 19 (5404+ / 2250-)
**Services Affected**: Backend (Django)

Solid feature with good domain modeling, comprehensive tests, and thoughtful patterns (consumer tracking, warehouse validation, dashboard cleanup). However, there are structural violations against the CLAUDE.md architecture, a SQL injection vector, and a RAG threshold logic bug that should be addressed before merge.

### What's Done Well
- Comprehensive test coverage — 4 test files covering API and service layers with edge cases (~1160 lines of tests)
- Clean model design — `Metric` → `KPI` with `PROTECT` on delete, `UniqueConstraint` on `(org, name)`
- Metric consumer tracking before delete prevents orphaned references
- Dashboard cleanup on KPI delete prevents broken layouts
- Warehouse validation on create/update catches bad definitions early
- Two metric definition paths (simple vs expression) — good UX for both simple and advanced users

See inline comments for specific blocking issues, suggestions, and nitpicks.

File: ddpui/core/metric/metric_service.py (hunk 1 above)
Comment (Ishankoradia) [reply ddpui/core/metric/metric_service.py:1]: Done

File: ddpui/services/metric_service.py (hunk 4 above)
Comment (Ishankoradia) [reply ddpui/services/metric_service.py]: Done

File: ddpui/services/metric_service.py (hunk 3 above)
Comment (Ishankoradia) [reply ddpui/services/metric_service.py]: This is done .

File: ddpui/models/metric.py (hunk 15 above)
Comment (Ishankoradia) [reply ddpui/models/metric.py:64]: Done

File: ddpui/api/kpi_api.py (hunk 16 above)
Comment (Ishankoradia) [reply ddpui/api/kpi_api.py]: Done
